File: python/NEAT/genome_history.py
import numpy as np
from .gene import *
from .node import *


class GenomeHistory(object):
    def __init__(self, inputs, outputs) -> None:
        self.inputs = inputs
        self.outputs = outputs

        self.all_genes: list[Gene] = []

        self.global_innovation = 0

        self.highest_hidden = 2

    def exists(self, n1: Node, n2: Node):
        for c in self.all_genes:
            if c.in_node.number == n1.number and c.out_node.number == n2.number:
                return c.clone()

        return None

File: treejs_test/neat/genome_history.ts
import { Gene } from "./gene.ts";
import { Node } from "./node.ts";

export class GenomeHistory {
  all_genes: Gene[] = []
  global_innovation = 0
  highest_hidden = 2

  constructor(
    public inputs: number,
    public outputs: number
  ) { }

  public exists(n1: Node, n2: Node) {
    const gene = this.all_genes.find((gene) =>
      gene.in_node.number === n1.number &&
      gene.out_node.number === n2.number
    )
    
    return gene ? gene.clone() : null
  }

  public debug_info() {
    console.log("[GenomeHistory] Size Genes: ", this.all_genes.length)
  }
}